### PR TITLE
Consume KVM devices using device plugin framework

### PR DESCRIPTION
**Pull request by stu-gott**
Incomplete implementation of KVM devices implemented as a Kubernetes Device Plugin. Submitted now in case reviewers would like to give early feedback.

- [x] Annotate KVM capable nodes
- [x] Implement server side KVM Device Plugins (inside virt-handler)
- [x] Implement server side tun Device Plugin
- [x] Ensure only as many DevicePlugin KVM objects as needed are created
- [x] Consume kvm devices from inside virt-launcher
- [x] Add scheduling constraint to pods if softwareEmulation isn't allowed
- [x] Drop permissions of virt-launcher
- [x] Unit tests
- [x] Functional tests
- [x] Update user guide

### CODE REVIEW

**Comment by vladikr** [comment]
:+1: Looks great so far

**Comment by stu-gott** [comment]
@fabiand yes I did, thank you. That was an excellent starting point.

**Comment by fabiand** [comment]
@stu-gott did you consider to at least have the DP in it's own container?
same pod (handler) sounds good.

**Comment by rmohr** [comment]
> @stu-gott did you consider to at least have the DP in it's own container?
> same pod (handler) sounds good.

Hm, that would make it again harder for the built-in health-check of virt-handler to detect if components of it are not working properly, to make it stop announcing that vms can be scheduled.

**Comment by fabiand** [comment]
> Hm, that would make it again harder for the built-in health-check of virt-handler to detect if components of it are not working properly, to make it stop announcing that vms can be scheduled.

So, the handler could also monitor the device plugin.

I'm fine with any approcah right now.

But at some point we might want to look why/if the handler needs to be privileged and then consider moving stuff to a DP.

File: cmd/virt-handler/virt-handler.go [hunk 1]
```diff
@@ -152,13 +157,19 @@ func (app *virtHandlerApp) AddFlags() {
...
 	flag.StringVar(&app.VirtShareDir, "kubevirt-share-dir", virtShareDir,
 		"Shared directory between virt-handler and virt-launcher")
 
 	flag.DurationVar(&app.WatchdogTimeoutDuration, "watchdog-timeout", defaultWatchdogTimeout,
 		"Watchdog file timeout")
+
+	// TODO: the Device Plugin API does not allow for infinitely available (shared) devices
+	// so the current approach is to register an arbitrary number.
+	// This should be deprecated if the API allows for shared resources in the future
+	flag.IntVar(&app.MaxDevices, "max-devices", maxDevices,
```
Comment (rmohr) [inline cmd/virt-handler/virt-handler.go]: Just hardcode it to e.g. 1000?

File: pkg/virt-handler/device-manager/generic_device.go [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+	"os"
+	"path"
+	"strconv"
+	"time"
+	"path/filepath"
+
+	"golang.org/x/net/context"
+
+	"google.golang.org/grpc"
+	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1alpha"
```
Comment (rmohr) [inline pkg/virt-handler/device-manager/generic_device.go]: That should be the only dependecy you need: `k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1beta1`. Also make sure that it aligns with the version which we use for client-go, currently `kubernetes-1.9.7`.

Comment (rmohr) [reply pkg/virt-handler/device-manager/generic_device.go]: Something like

```diff
diff --git a/glide.yaml b/glide.yaml
index 784b28df..61ff8ee0 100644
--- a/glide.yaml
+++ b/glide.yaml
@@ -52,6 +52,10 @@ import:
   version: ^0.0.2
   subpackages:
   - cobra
+- package: k8s.io/kubernetes
+  version: kubernetes-1.9.7
+  subpackages:
+  - /pkg/kubelet/apis/deviceplugin/v1beta1
 - package: k8s.io/apiextensions-apiserver
   version: kubernetes-1.9.7
 - package: github.com/ant31/crd-validation
```

should do it

Comment (rmohr) [reply pkg/virt-handler/device-manager/generic_device.go]: Edit: If you use `v1.10.0` you can use `v1beta`. For `v1.9.7` you will have to use `v1alpha1`.

Comment (stu-gott) [reply pkg/virt-handler/device-manager/generic_device.go]: Actually we should merge this after Kubernetes 1.10.X is used as a provider. Requiring a feature-gate to use an alpha feature (in 1.9.X) is suboptimal.

File: cmd/virt-handler/virt-handler.go (hunk 1 above)
Comment (stu-gott) [reply cmd/virt-handler/virt-handler.go]: Outcome of a discussion with @fabiand was to make this configurable. However since the medium term goal would be to get rid of the need for this setting, the notion of making a magic number isn't terribly disagreeable.

There are memory and consumption side effects to making this number too large. I'd suggest the Kubelet default of 110 is a better choice (we can't realistically need more devices than there are available pods anyways).

File: pkg/virt-handler/device-manager/generic_device.go (hunk 2 above)
Comment (stu-gott) [reply pkg/virt-handler/device-manager/generic_device.go]: This PR is probably gated on: https://github.com/kubevirt/kubevirt/pull/1145

Comment (rmohr) [reply pkg/virt-handler/device-manager/generic_device.go]: We have k8s 1.10.3 as default for some time already

**Comment by rmohr** [comment]
> @stu-gott if glide insists on pulling in complete kubernetes and not just these two files, then simply copy them. It is an API definition anyway.

@stu-gott with copying I literally meant copying the required files to our code-base. Now anyone who would run `make dep-install` would get the whole k8s tree.

**Comment by stu-gott** [comment]
@rmohr I did not request the kubernetes code tree in glide.yaml. I'm not sure what's triggering that to be pulled in.

**Comment by stu-gott** [comment]
@rmohr deps are fixed. I think this PR is ready for review.

**Comment by stu-gott** [comment]
retest this please

**Comment by fabiand** [comment]
Beautiful!

Beautiful?

/cc @vladikr

**Comment by vladikr** [comment]
Great work Stu!

Looks good to me overall. 

There are some parts that I didn't quite understand, like the use of response.ContainerResponses in Allocate().
Too bad it's impossible to have dependant PRs :) - It was hard to follow such a large work.

File: pkg/virt-launcher/virtwrap/cmd-server/server.go [hunk 3]
```diff
@@ -63,7 +63,7 @@ func (s *Launcher) Sync(args *cmdclient.Args, reply *cmdclient.Reply) error {
 		return nil
 	}
 
-	_, err = s.domainManager.SyncVMI(vmi, s.allowEmulation)
+	_, err = s.domainManager.SyncVMI(vmi, s.useEmulation)
```
Comment (davidvossel) [inline pkg/virt-launcher/virtwrap/cmd-server/server.go]: can't we determine if useEmulation should be used or not based on the vmi requests field?  It doesn't seem like the cmd server needs to know about useEmulation before hand

Comment (davidvossel) [reply pkg/virt-launcher/virtwrap/cmd-server/server.go]: > I think we were specifically trying to avoid any reference to hardware acceleration inside the VM/VMI specs, hence the cluster-wide setting.

is there a way to detect within the virt-launcher pod whether the kvm device is present or not?

**Comment by fabiand** [comment]
No. The controller decides to either add or not add the kvm device.

The operator can configure the cluster to use emulation if no hardware
acceleration is available.
This was the agreement we came to.

File: pkg/virt-launcher/virtwrap/cmd-server/server.go (hunk 3 above)
Comment (fabiand) [reply pkg/virt-launcher/virtwrap/cmd-server/server.go]: No. The controller - cluster level - decides to either add or not add the kvm device. The operator can configure the cluster to use emulation if no hardware acceleration is available. This was the agreement we came to.

File: pkg/virt-handler/device-manager/device_controller.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+			return err
+		}
+	} else {
+		// File already exists, so there's nothing to wait for
+		return nil
+	}
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil
```
Comment (cynepco3hahue) [inline pkg/virt-handler/device-manager/device_controller.go]: `return err`

File: pkg/virt-handler/device-manager/deviceplugin/v1alpha/BUILD [hunk 5]
```diff
@@ -0,0 +1,41 @@
+package(default_visibility = ["//visibility:public"])
```
Comment (cynepco3hahue) [inline pkg/virt-handler/device-manager/deviceplugin/v1alpha/BUILD]: I just curious why do we have all this files(`pkg/virt-handler/device-manager/deviceplugin`), can we just import them without copy paste?

File: pkg/virt-launcher/virtwrap/cmd-server/server.go (hunk 3 above)
Comment (stu-gott) [reply pkg/virt-launcher/virtwrap/cmd-server/server.go]: @davidvossel a little context: this is a side effect of the scheduling contract with device plugins. Basically a node advertises "yes. I have KVM devices" (these show up as namespaced entities in the `allocatable` and `capacity` sections of the node's yaml. e.g. `devices.kubevirt.io/kvm: "110"`). Thus when virt-controller makes the decision to add a request for a KVM device to a pod, that pod can now *only* be scheduled on nodes which provide KVM devices.

Or stated another way, the decision to request a KVM device basically mandates that one will be present (or the other outcome is that the pod will hang out in scheduling limbo forever). This motivated the change of semantics from "allowEmulation" to "useEmulation" as the virt-controller can no longer take a passive role in the process.

**Comment by cynepco3hahue** [comment]
We need to decide what we are going to do with CI because currently all `1.9` and `3.9` jobs fail on a VM creation.
One of the possibilities, to check the version before running tests and create a relevant config map with `useEmulation: true`

**Comment by stu-gott** [comment]
Full log of test suite run: https://paste.fedoraproject.org/paste/mOKaxL6G-ueROpHYFDcG9g
2 tests show up as failing. so I retested them individually.

storage test passes on second attempt. https://paste.fedoraproject.org/paste/Y8zhi0b6K4OWLWZ1jNuoIQ
(presumed flakey)

networking functest fails several times in a row, but something's fishy there: https://paste.fedoraproject.org/paste/MDODpYemdcPARF3i0NxKiQ  ... it runs the (focused) test twice it seems. once it passes, and once it fails (during the same run)

pod logs captured during the test run look fine... https://paste.fedoraproject.org/paste/VhJDhlCghAVv9qT4JQGJ3A

**Comment by fabiand** [comment]
@cynepco3hahue we could also drop 1.9 and 3.9 support. and instead add k8s and origin master support?

File: pkg/virt-handler/device-manager/deviceplugin/v1alpha/BUILD (hunk 5 above)
Comment (stu-gott) [reply pkg/virt-handler/device-manager/deviceplugin/v1alpha/BUILD]: That pulled in the entire kubernetes tree as a dependency, which means we would have had 9500+ files added to the vendor tree for no good reason. I assume this will be sorted out upstream at some point as it affects all consumers of device plugins.

**Comment by stu-gott** [comment]
Latest commit fixes the error @cynepco3hahue noted with respect to returning `nil` incorrectly. This will not affect the above test suite runs as testing that `fsnotify.NewWatcher()` fails is not tested for (or easy to test or reasonable to do anyways)

**Comment by fabiand** [comment]
Sounds like the state is good enough.

Merging it, let's give it some time to see the fallout.

**Comment by cynepco3hahue** [comment]
> Full log of test suite run: https://paste.fedoraproject.org/paste/mOKaxL6G-ueROpHYFDcG9g
> 2 tests show up as failing. so I retested them individually.

I am curious why network tests skipped
```
Skip network test that requires multiple nodes when only one node is present.
```
All our environments must have two nodes.

File: pkg/virt-launcher/virtwrap/cmd-server/server.go (hunk 3 above)
Comment (davidvossel) [reply pkg/virt-launcher/virtwrap/cmd-server/server.go]: works for me :+1:

**Comment by fabiand** [comment]
@stu-gott this PR is worth a blog post!
